Plugin to display the open files upon test failure

jbarnoud opened the pull request:
Add a plugin that list open files when a test fails. This aims at testing the hypothesis that we have to many opened file descriptors, and that it could cause #853.

The plugin is enabled by default. It displays the path to the file that
have a open file handler. To disable the plugin, run mda_nosetests with
the --no-open-files option.

The commit also introduces a test that fails if the MDA_FAILURE_TEST
environment variable is set. This test comes handy to test nose behavior
upon failure.

The open-files plugin requires psutil. That requirement has been added
to setup.py.

kain88-de commented:
could you also activate the extension then we see straight if it works.

kain88-de commented:
I just see it is. Boy do we have a lot of files open at the same time. Doesn't look like deleting a universe closes the files associated with it.

@jbarnoud it would be nice to have a count of open files at the beginning of the output.

kain88-de commented:
considering that we always have a few files open I thought I tried out something.

```python
import os
import psutil
import MDAnalysis as mda
from MDAnalysisTests.datafiles import PSF, DCD, PDB, XTC

u = mda.Universe(PSF, DCD)
u = mda.Universe(PDB, XTC)
del u

p = psutil.Process(os.getpid())
for f in p.open_files():
    print f.path 
```

This gives the following output.

```
/home/max/.ipython/profile_default/history.sqlite
/home/max/.ipython/profile_default/history.sqlite
/home/max/conda/lib/python2.7/site-packages/MDAnalysisTests/data/adk_dims.dcd
/home/max/conda/lib/python2.7/site-packages/MDAnalysisTests/data/adk_oplsaa.xtc
/home/max/.ipython/profile_default/history.sqlite-journal
```

The files are still open even though the universe should be gone. When I call the readers directly and delete/close them the open file handle is gone. 

It doesn't even help to manually call the garbage collector some times. This is weird the direct garbage collection sometimes helps but not all the times.

kain88-de commented:
OK after playing around with adding `__del__` methods and prints inside them I come to believe we have a problem with the cyclic references in the atomgroups of a universe.  

```
>>>  import sys
>>> import MDAnalysis as mda
>>> from MDAnalysisTests.datafiles import DCD, PSF

>>> u = mda.Universe(PSF, DCD)
>>> dcd = mda.coordinates.DCD.DCDReader(DCD)

>>> sys.getrefcount(u)
47683
>>> sys.getrefcount(dcd)
2
```

why is the reference count of a universe object so crazy high? Is it because **every** single Atom object stores a reference to the universe it comes from?

richardjgowers commented:
Yeah each Atom has a pointer to Universe for accessing the trajectory

On Sun, 12 Jun 2016, 13:56 kain88-de, <notifications@github.com> wrote:

> OK after playing around with adding __del__ methods and prints inside
> them I come to believe we have a problem with the cyclic references in the
> atomgroups of a universe.
>
> >>>  import sys
> >>> import MDAnalysis as mda
> >>> from MDAnalysisTests.datafiles import DCD, PSF
>
> >>> u = mda.Universe(PSF, DCD)
> >>> dcd = mda.coordinates.DCD.DCDReader(DCD)
>
> >>> sys.getrefcount(u)
> 47683
> >>> sys.getrefcount(dcd)
> 2
>
> why is the reference count of a universe object so crazy high? Is it
> because *every* single Atom object stores a reference to the universe it
> comes from?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/874#issuecomment-225430909>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AI0jB8CqukcUEuiufmS2PZ6BR8DaYkI-ks5qLAIDgaJpZM4IzFwH>
> .
>

kain88-de commented on testsuite/MDAnalysisTests/plugins/open_files.py:
```diff
@@ -0,0 +1,79 @@
+        handle_couter = collections.Counter(open_files)
+        new_ev = u'\n'.join(
+            [ev]
+            + [ln(u'>>> There {} open file handlers for {} files: <<<'
```
Please also print the number of known universe objects in `MDAnalysis._anchor_universe`. This helps to decide if we keep universe objects alive after they should have been destroyed if the number of universe objects is roughly the same as the number of open files.

kain88-de replied:
You can even iterate over the set and count the different reader types to check if they match up with the numbers of open filetypes.

jbarnoud replied:
I'll do that. For now I try to fix a bug in the plugin: the list of open file is not the list of open handles at the end of the test but at the end of the test suite.

kain88-de commented on testsuite/MDAnalysisTests/plugins/open_files.py:
```diff
@@ -86,24 +86,25 @@ def formatFailure(self, test, err):
-        handlers = process.open_files()
-        path = [f.path for f in handlers]
-        return path
+def _gather_open_files(self):
```
you need to remove the `self` argument.

jbarnoud replied:
How can you be that fast!? The extra `self` stayed there for less than 5 
minutes!!!

On 13/06/16 09:45, kain88-de wrote:
>
> In testsuite/MDAnalysisTests/plugins/open_files.py 
> <https://github.com/MDAnalysis/mdanalysis/pull/874#discussion_r66750705>:
>
> >
> > -        If several file handle are open for a given file, then the path will
> > -        appear as many time as they are open file handle for that file.
> > -        """
> > -        process = psutil.Process(os.getpid())
> > -        handlers = process.open_files()
> > -        path = [f.path for f in handlers]
> > -        return path
> > +def _gather_open_files(self):
>
> you need to remove the |self| argument.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/MDAnalysis/mdanalysis/pull/874/files/5603421ec8ecc5045a844088c8b284ee82c50806..64f0143ff39a691016cdd0f05d25c4a32673e3cc#r66750705>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe/ABUWuiox6jeDzyb8Kh13mN-dqUJzghHqks5qLQqEgaJpZM4IzFwH>.
>

kain88-de replied:
lucky coincidence

kain88-de replied:
and the coveralls bot notified me that there are changes.

jbarnoud commented:
I updated the CHANGELOG and rebased against develop. The PR is now ready for review.

@kain88-de I prepare a new PR with an other plugin to follow universes, it makes sense to me to have these plugins separated.

kain88-de commented on testsuite/MDAnalysisTests/test_failure.py:
```diff
@@ -0,0 +1,26 @@
+
+
+def test_failure():
+    """Fail if the MDA_FAILURE_TEST environment variable is set.
```
what is this variable and why is this test included now?

jbarnoud replied:
On 13/06/16 12:19, kain88-de wrote:
>
> In testsuite/MDAnalysisTests/test_failure.py 
> <https://github.com/MDAnalysis/mdanalysis/pull/874#discussion_r66769392>:
>
> > +# Oliver Beckstein and contributors (see AUTHORS for the full list)
> > +#
> > +# Released under the GNU Public Licence, v2 or any higher version
> > +#
> > +# Please cite your use of MDAnalysis in published work:
> > +#
> > +# N. Michaud-Agrawal, E. J. Denning, T. B. Woolf, and O. Beckstein.
> > +# MDAnalysis: A Toolkit for the Analysis of Molecular Dynamics Simulations.
> > +# J. Comput. Chem. 32 (2011), 2319--2327, doi:10.1002/jcc.21787
> > +#
> > +
> > +import os
> > +
> > +
> > +def test_failure():
> > +    """Fail if the MDA_FAILURE_TEST environment variable is set.
>
> what is this variable and why is this test included now?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/MDAnalysis/mdanalysis/pull/874/files/50bd3ea09e9ceaaae4b2eea4856c20612ac85d14#r66769392>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe/ABUWus0BwcwjKBTPxQNUN2OgdOnUgRlGks5qLS6wgaJpZM4IzFwH>.
>
This test allows to have at least one failing test on command. Just set 
the MDA_FAILURE_TEST environment variable and this test will fail. 
Thanks to that very useful tool, you do not have to sabotage a test to 
see if your plugin works on test failures.

kain88-de replied:
could you document that as well. I would like to have it in the README so I can access the information offline as well. But feel free to document it where ever you feel it appropriate.

It should definitely be in the CHANGELOG as well.

jbarnoud commented:
@kain88-de I updated the CHANGELOG to account for `test_failure`. I'll add some doc in the [wiki page on testing](https://github.com/MDAnalysis/mdanalysis/wiki/UnitTests) once the PR merged.
